crates/phactory/src/prpc_service.rs

```diff
@@ -1714,8 +1905,60 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi for Rpc
             .statistics(request)
             .map_err(from_debug)
     }
-}
-
-fn try_decode_hex(hex_str: &str) -> Result<Vec<u8>, hex::FromHexError> {
-    hex::decode(hex_str.strip_prefix("0x").unwrap_or(hex_str))
+    async fn generate_cluster_state_request(
+        &mut self,
+        _: (),
+    ) -> Result<pb::SaveClusterStateArguments, prpc::server::Error> {
+        let mut phactory = self.lock_phactory(true, false)?;
+        let system = phactory.system()?;
+        if system.contract_cluster.is_some() {
```

Comment:
Why it would make DoS harder? The attacker can always sign a fake request without pruntime and the server side would take exactly the same cost to validate the request.

Reply:
I think the DoS vector is someone create a request without actually being added to a cluster. In that case, anyone can generate such request with minimum cost, and ask the old workers to dump the request. The dump is heavy. So the old workers can get DoS-ed.

Reply:
The old worker always ensures the receiver worker is in the cluster before dumping the state.

Reply:
OK. I see it's verified.